### DISCUSSION TITLE
Add metadata to events

### DISCUSSION THREAD

**Discussion by peteglassey** [Eventuous/eventuous#283] [General]
I cant seem to find it in the documentation, but how do we add metadata to our events that are going to eventstoredb? Eg i want to decorate each aggregate's event metadata with some additional information like version, content-type, scope (internal construct). 

I thought it would be on the Apply() method on the Aggregate<T> base class, but isnt. 

In our aggregates we have command handling methods, that create the event and call Apply(), so i thought there might be an overload, but i cant find it.

I see there is a way to append metadata when publishing an event to a publisher (eg an 'integration' event etc), but nothing i can find when appending to aggregate's stream.

What did i miss?

**Comment by alexeyzimarev** [comment]
It's supported, but it's pretty weak in terms of the API. `AggregateStore`'s constructor accepts a function, which allows you to add metadata, or do something else with the event that's going to be persisted. You would need to inspect the proposed `StreamEvent` and understand what you want to add in the meta. If you only want to add something generic like a timestamp or diagnostic data, then it will work just fine.

```
    public AggregateStore(
        IEventReader                    reader,
        IEventWriter                    writer,
        Func<StreamEvent, StreamEvent>? amendEvent      = null,
        AggregateFactoryRegistry?       factoryRegistry = null
    )
```

**Reply by peteglassey** [reply]
Thanks Alexy. Ive copied your Services.AddAggregateStore<EsdbEventStore>() method and extended it to provide a factory method for the DI container to add the event metadata from a custom event attribute we have.

    public static IServiceCollection AddAggregateStoreX<T>(this IServiceCollection services)
        where T : class, IEventStore {
        services.TryAddSingleton<AggregateFactoryRegistry>();

        if (EventuousDiagnostics.Enabled) {
            services
                .AddSingleton<T>()
                .AddSingleton(sp => TracedEventStore.Trace(sp.GetRequiredService<T>()));
        }
        else {
            services.AddSingleton<IEventStore, T>();
        }

        services.AddSingleton<IAggregateStore>(provider =>
        {
            return new AggregateStore(
                provider.GetRequiredService<IEventStore>(),
                provider.GetRequiredService<IEventStore>(),
                streamEvent =>
                {
                    var eventTypeAttr = GetEventAttribute(streamEvent.Payload);
                    
                    streamEvent.Metadata.Add(PacificMetadata.MetadataKeys.Scope, eventTypeAttr.Scope);
                    streamEvent.Metadata.Add(PacificMetadata.MetadataKeys.Encoding, PacificMetadata.ContentTypes.Json);
                    streamEvent.Metadata.Add(PacificMetadata.MetadataKeys.Version, eventTypeAttr.Version);
                    return streamEvent;
                });
        });
        
        return services;
        
        Pacific.Events.Library.Attributes.EventTypeAttribute GetEventAttribute(object? streamEventPayload)
        {
            if (streamEventPayload is null) throw new InvalidOperationException("Cannot get version metadata of a null event!");
            
            return streamEventPayload.GetType().GetCustomAttributes(typeof(Pacific.Events.Library.Attributes.EventTypeAttribute), true)
                .FirstOrDefault() is Pacific.Events.Library.Attributes.EventTypeAttribute attribute
                ? attribute
                : throw new Exception("Could not find EventTypeAttribute on event");
        }
    }

**Reply by alexeyzimarev** [reply]
I think it can be made easier and you would avoid a custom DI container registration function, if `AggregateStore` would accept a named delegate like `AmendAggregateStoreEvent` instead of `Func<StreamEvent, StreamEvent>`, so the function can be registered and resolved normally.

**Reply by peteglassey** [reply]
Hi Alexey, yes, i agree it could be made easier. Perhaps an IMetadataProvider interface in the ctor of Aggregate store, and you can register a default one (which no-op), but people can register their own as/when required?

**Reply by alexeyzimarev** [reply]
An interface is not in any way better compared to named delegate, I generally avoid interfaces with one function in it, especially when it's called a `Provider` :) You can register functions just as you register other dependencies. I'd say it should be nullable, so if there's nothing provided there will be no call.

**Reply by peteglassey** [reply]
Alexey - are you intending to add this feature any time soon?

**Reply by alexeyzimarev** [reply]
It's already there https://github.com/Eventuous/eventuous/pull/285

**Reply by peteglassey** [reply]
Thanks heaps

**Reply by alexeyzimarev** [reply]
It has never made to MyGet because of the failing tests in the pipeline (they actually pass but the container fails to report its health). I will hopefully fix it soon enough.
